File: lib/keyTracker.py
"""run using this command: 
python .\lib\keyTracker.py"""
from __future__ import annotations
import os
import signal
import sys
import json
from collections import Counter, deque
from contextlib import contextmanager
from datetime import datetime
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parents[1]
OUTPUT_FOLDER = PROJECT_ROOT / "key-tracker-js-output"
LOG_FILE = OUTPUT_FOLDER / f"key_presses_{datetime.now():%Y%m%d_%H%M%S_%f}.js"
DEFAULT_WAIT_MS = 500
ACTION_INDENT = "            "
STEP_BLOCK_INDENT = "          "
STEP_INDENT = "    "
STEPS_INDENT = "      "

KEY_NAMES = {
    "UP": "Up",
    "DOWN": "Down",
    "LEFT": "Left",
    "RIGHT": "Right",
    "HOME": "Home",
    "ENTER": "Enter",
}

SHORTCUT_KEYS = {
    "w": "WAIT",
    "f": "FINISH",
    "i": "SET_INTERVAL",
    "t": "NEXT_TOPIC",
    "p": "CAPTURE_REUSE_PREVIOUS",
    "r": "CAPTURE_REUSE_1_3",
    "q": "QUIT",
}

CAPTURE_REUSE_IMAGES = {
    "CAPTURE_REUSE_PREVIOUS": "previous",
    "CAPTURE_REUSE_1_3": "1-3",
}

Action = tuple[str, str | int]


class KeyReader:
    """Read single key presses on Windows and POSIX terminals."""

    def __init__(self) -> None:
        self.is_windows = os.name == "nt"

    @contextmanager
    def raw_mode(self):
        if self.is_windows:
            yield
            return

        import termios
        import tty

        fd = sys.stdin.fileno()
        original_settings = termios.tcgetattr(fd)
        try:
            tty.setcbreak(fd)
            yield
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, original_settings)

    def read_key(self) -> str | None:
        if self.is_windows:
            return self._read_windows_key()

        return self._read_posix_key()

    def _read_windows_key(self) -> str | None:
        import msvcrt

        key = msvcrt.getwch()

        if key in ("\x00", "\xe0"):
            arrow = msvcrt.getwch()
            return {
                "H": "UP",
                "P": "DOWN",
                "K": "LEFT",
                "M": "RIGHT",
            }.get(arrow)

        if key == "\r":
            return "ENTER"
        if key in ("\b", "\x7f"):
            return "UNDO"
        if key == " ":
            return "HOME"

        return SHORTCUT_KEYS.get(key.lower())

    def _read_posix_key(self) -> str | None:
        key = sys.stdin.read(1)

        if key == "\x1b":
            sequence = key + sys.stdin.read(2)
            if sequence == "\x1b[A":
                return "UP"
            if sequence == "\x1b[B":
                return "DOWN"
            if sequence == "\x1b[D":
                return "LEFT"
            if sequence == "\x1b[C":
                return "RIGHT"
            return None

        if key in ("\n", "\r"):
            return "ENTER"
        if key in ("\b", "\x7f"):
            return "UNDO"
        if key == " ":
            return "HOME"

        return SHORTCUT_KEYS.get(key.lower())


def clear_screen() -> None:
    print("\033[2J\033[H", end="")


def prompt_topic() -> str:
    clear_screen()
    print("Topic")
    print("=====")
    return input("Enter topic #: ").strip()


def prompt_wait_interval() -> int:
    while True:
        answer = input("Set the interval: ").strip()
        try:
            interval = int(answer)
        except ValueError:
            interval = 0

        if interval > 0:
            return interval

        print("Interval must be a positive whole number.")


def initialize_log_file() -> None:
    OUTPUT_FOLDER.mkdir(parents=True, exist_ok=True)
    with LOG_FILE.open("w", encoding="utf-8") as log_file:
        log_file.write("export const keyPresses = {\n")


def log_file_display_path() -> str:
    try:
        return str(LOG_FILE.relative_to(PROJECT_ROOT))
    except ValueError:
        return str(LOG_FILE)


def finalize_log_file() -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write("};\n")
        log_file.write("\nexport default keyPresses;\n")


def start_topic(topic: str) -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(f"{STEP_INDENT}{json.dumps(topic)}: {{\n")
        log_file.write(f"{STEPS_INDENT}steps: [\n")


def start_step() -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(f"{STEP_BLOCK_INDENT}[\n")


def log_key_press(key: str) -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(f'{ACTION_INDENT}{{ type: "remote", key: "KEY_{key}" }},\n')


def log_wait(wait_ms: int) -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(f'{ACTION_INDENT}{{ type: "wait", ms: {wait_ms} }},\n')


def log_capture_reuse(reuse_image: str) -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(
            f'{ACTION_INDENT}{{ type: "capture", mode: "reuse", reuseImage: "{reuse_image}" }},\n'
        )


def finish_step() -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(f"{STEP_BLOCK_INDENT}],\n")


def finish_topic() -> None:
    with LOG_FILE.open("a", encoding="utf-8") as log_file:
        log_file.write(f"{STEPS_INDENT}],\n")
        log_file.write(f"{STEP_INDENT}}},\n")


def last_content_line_index(lines: list[str]) -> int:
    index = len(lines) - 1
    while index >= 0 and lines[index] == "":
        index -= 1
    return index


def write_log_lines(lines: list[str]) -> None:
    LOG_FILE.write_text("\n".join(lines), encoding="utf-8")


def remove_empty_open_step() -> bool:
    lines = LOG_FILE.read_text(encoding="utf-8").split("\n")
    index = last_content_line_index(lines)
    if index < 0 or lines[index] != f"{STEP_BLOCK_INDENT}[":
        return False

    del lines[index]
    write_log_lines(lines)
    return True


def remove_last_step_close() -> bool:
    lines = LOG_FILE.read_text(encoding="utf-8").split("\n")
    index = last_content_line_index(lines)
    if index < 0 or lines[index] != f"{STEP_BLOCK_INDENT}],":
        return False

    del lines[index]
    write_log_lines(lines)
    return True


def remove_last_action_line() -> bool:
    lines = LOG_FILE.read_text(encoding="utf-8").split("\n")

    for index in range(last_content_line_index(lines), -1, -1):
        if lines[index].startswith(ACTION_INDENT):
            del lines[index]
            write_log_lines(lines)
            return True

        if lines[index] == f"{STEP_BLOCK_INDENT}[":
            break

    return False


def find_last_remote_key(actions: list[Action]) -> str | None:
    for action_type, value in reversed(actions):
        if action_type == "remote":
            return str(value)

    return None


def recompute_counts(counts: Counter[str], actions: list[Action]) -> str | None:
    counts.clear()
    for action_type, value in actions:
        if action_type == "remote":
            counts[str(value)] += 1

    return find_last_remote_key(actions)


def describe_action(action: Action) -> str:
    action_type, value = action
    if action_type == "wait":
        return f"Wait {value} ms"
    if action_type == "capture":
        return f"Capture reuse {value}"

    return KEY_NAMES[str(value)]


def draw(
    counts: Counter[str],
    recent: deque[str],
    last_key: str | None,
    topic_open: bool,
    step_open: bool,
    current_topic: str | None,
    wait_ms: int,
) -> None:
    clear_screen()
    print("W to add a wait")
    print("F to finish current step.")
    print("I to set wait interval.")
    print("T to start the next topic.")
    print("P to capture previous image.")
    print('R to reuse images "1-3".')
    print("Backspace to undo the last action.")
    print("Q to end program.\n")
    print(f"Saving presses to: {log_file_display_path()}")
    print(f"Wait interval: {wait_ms} ms")
    if topic_open and step_open and current_topic:
        print(f"Current topic: {current_topic}\n")
    elif current_topic:
        print(f"Current topic: {current_topic} (ready)\n")
    else:
        print("Current topic: not started\n")

    for key in ("UP", "DOWN", "LEFT", "RIGHT", "HOME", "ENTER"):
        print(f"{KEY_NAMES[key]}: {counts[key]}")

    if last_key:
        print(f"\nLast key: {KEY_NAMES[last_key]}")
    else:
        print("\nLast key: none yet")

    if recent:
        print("\nKey Counter:")
        for event in recent:
            print(f"  {event}")


def main() -> int:
    original_sigint_handler = signal.getsignal(signal.SIGINT)
    signal.signal(signal.SIGINT, signal.SIG_IGN)

    reader = KeyReader()
    counts: Counter[str] = Counter()
    recent: deque[str] = deque(maxlen=3)
    last_key: str | None = None
    topic_open = False
    step_open = False
    current_topic: str | None = None
    current_step_actions: list[Action] = []
    finished_step_actions: list[list[Action]] = []
    wait_ms = DEFAULT_WAIT_MS

    def open_step_if_needed() -> None:
        nonlocal current_topic, current_step_actions, step_open, topic_open

        if step_open:
            return

        if current_topic is None:
            current_topic = prompt_topic()
        if not topic_open:
            start_topic(current_topic)
            topic_open = True
        start_step()
        current_step_actions = []
        step_open = True

    def redraw() -> None:
        draw(counts, recent, last_key, topic_open, step_open, current_topic, wait_ms)

    initialize_log_file()

    try:
        current_topic = prompt_topic()
        redraw()

        while True:
            with reader.raw_mode():
                key = reader.read_key()

            if key == "QUIT":
                break
            timestamp = datetime.now()

            if key == "SET_INTERVAL":
                wait_ms = prompt_wait_interval()
                recent.appendleft(f"{timestamp:%H:%M:%S} - Wait interval {wait_ms} ms")
                redraw()
                continue

            if key == "WAIT":
                open_step_if_needed()
                log_wait(wait_ms)
                current_step_actions.append(("wait", wait_ms))
                finished_step_actions.append(current_step_actions.copy())
                finish_step()
                current_step_actions = []
                step_open = False
                counts.clear()
                last_key = None
                recent.appendleft(
                    f"{timestamp:%H:%M:%S} - Wait {wait_ms} ms; finished step"
                )
                redraw()
                continue

            if key == "FINISH":
                if step_open:
                    if current_step_actions:
                        finished_step_actions.append(current_step_actions.copy())
                        finish_step()
                    else:
                        remove_empty_open_step()
                    current_step_actions = []
                    step_open = False
                    recent.appendleft(f"{timestamp:%H:%M:%S} - Finished step")
                    counts.clear()
                    last_key = None
                else:
                    recent.appendleft(f"{timestamp:%H:%M:%S} - No open step to finish")
                redraw()
                continue

            if key == "UNDO":
                if not step_open and topic_open and finished_step_actions:
                    if remove_last_step_close():
                        current_step_actions = finished_step_actions.pop()
                        last_key = recompute_counts(counts, current_step_actions)
                        step_open = True
                    else:
                        recent.appendleft(
                            f"{timestamp:%H:%M:%S} - Could not go back to step"
                        )
                        redraw()
                        continue

                if not step_open or not current_step_actions:
                    recent.appendleft(f"{timestamp:%H:%M:%S} - Nothing to undo")
                    redraw()
                    continue

                action = current_step_actions.pop()
                if not remove_last_action_line():
                    current_step_actions.append(action)
                    recent.appendleft(
                        f"{timestamp:%H:%M:%S} - Could not undo last action"
                    )
                    redraw()
                    continue

                if action[0] == "remote":
                    counts[str(action[1])] = max(0, counts[str(action[1])] - 1)

                if current_step_actions:
                    last_key = find_last_remote_key(current_step_actions)
                else:
                    remove_empty_open_step()
                    step_open = False
                    counts.clear()
                    last_key = None

                recent.appendleft(
                    f"{timestamp:%H:%M:%S} - Removed {describe_action(action)}"
                )
                redraw()
                continue

            if key == "NEXT_TOPIC":
                if step_open:
                    if current_step_actions:
                        finished_step_actions.append(current_step_actions.copy())
                        finish_step()
                    else:
                        remove_empty_open_step()
                    current_step_actions = []
                    step_open = False
                if topic_open:
                    finish_topic()
                    topic_open = False

                counts.clear()
                last_key = None
                finished_step_actions.clear()
                current_topic = prompt_topic()
                recent.appendleft(
                    f"{datetime.now():%H:%M:%S} - Ready for topic {current_topic}"
                )
                redraw()
                continue

            if key in CAPTURE_REUSE_IMAGES:
                reuse_image = CAPTURE_REUSE_IMAGES[key]
                open_step_if_needed()
                log_capture_reuse(reuse_image)
                current_step_actions.append(("capture", reuse_image))
                recent.appendleft(
                    f"{timestamp:%H:%M:%S} - Capture reuse {reuse_image}"
                )
                redraw()
                continue

            if key not in KEY_NAMES:
                continue

            counts[key] += 1
            last_key = key
            open_step_if_needed()
            log_key_press(key)
            current_step_actions.append(("remote", key))
            recent.appendleft(f"{timestamp:%H:%M:%S} - {KEY_NAMES[key]}")
            redraw()
    except KeyboardInterrupt:
        pass
    finally:
        signal.signal(signal.SIGINT, original_sigint_handler)
        if step_open:
            if current_step_actions:
                finish_step()
            else:
                remove_empty_open_step()
        if topic_open:
            finish_topic()
        finalize_log_file()

    return 0

if __name__ == "__main__":
    raise SystemExit(main())

File: lib/keyTracker.mjs
/**
 * Interactive remote key / wait / capture logger (ported from lib/keyTracker.py).
 *
 * Run from repo root:
 *   node .\lib\keyTracker.mjs
 */
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import readline from "node:readline/promises";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = path.resolve(__dirname, "..");
const OUTPUT_FOLDER = path.join(PROJECT_ROOT, "key-tracker-js-output");

const DEFAULT_WAIT_MS = 500;
const ACTION_INDENT = "            ";
const STEP_BLOCK_INDENT = "          ";
const STEP_INDENT = "    ";
const STEPS_INDENT = "      ";

const KEY_NAMES = {
  UP: "Up",
  DOWN: "Down",
  LEFT: "Left",
  RIGHT: "Right",
  HOME: "Home",
  ENTER: "Enter",
};

const SHORTCUT_KEYS = {
  w: "WAIT",
  f: "FINISH",
  i: "SET_INTERVAL",
  t: "NEXT_TOPIC",
  p: "CAPTURE_REUSE_PREVIOUS",
  r: "CAPTURE_REUSE_1_3",
  q: "QUIT",
};

const CAPTURE_REUSE_IMAGES = {
  CAPTURE_REUSE_PREVIOUS: "previous",
  CAPTURE_REUSE_1_3: "1-3",
};

/** @typedef {[string, string | number]} Action */

let LOG_FILE = "";

function logStamp() {
  const d = new Date();
  const z2 = (n) => String(n).padStart(2, "0");
  const head = `${d.getFullYear()}${z2(d.getMonth() + 1)}${z2(d.getDate())}_${z2(d.getHours())}${z2(d.getMinutes())}${z2(d.getSeconds())}`;
  const tail = String(d.getMilliseconds() * 1000 + crypto.randomInt(1000)).padStart(6, "0");
  return `${head}_${tail}`;
}

function tryParseKey(buf) {
  if (buf.length === 0) return { needMore: true };

  const b0 = buf[0];

  if (b0 === 0x00 || b0 === 0xe0) {
    if (buf.length < 2) return { needMore: true };
    const arrows = { 0x48: "UP", 0x50: "DOWN", 0x4b: "LEFT", 0x4d: "RIGHT" };
    const k = arrows[buf[1]];
    if (!k) return { key: null, consume: 2 };
    return { key: k, consume: 2 };
  }

  if (b0 === 0x1b) {
    if (buf.length < 3) return { needMore: true };
    if (buf[1] === 0x5b) {
      const sub = { 0x41: "UP", 0x42: "DOWN", 0x44: "LEFT", 0x43: "RIGHT" }[buf[2]];
      if (sub) return { key: sub, consume: 3 };
    }
    return { key: null, consume: 3 };
  }

  if (b0 === 0x0d || b0 === 0x0a) return { key: "ENTER", consume: 1 };
  if (b0 === 0x08 || b0 === 0x7f) return { key: "UNDO", consume: 1 };
  if (b0 === 0x20) return { key: "HOME", consume: 1 };

  if (b0 < 0x80) {
    const ch = String.fromCharCode(b0);
    const shortcut = SHORTCUT_KEYS[ch.toLowerCase()];
    if (shortcut) return { key: shortcut, consume: 1 };
    return { key: null, consume: 1 };
  }

  let len = 1;
  if ((b0 & 0xe0) === 0xc0) len = 2;
  else if ((b0 & 0xf0) === 0xe0) len = 3;
  else if ((b0 & 0xf8) === 0xf0) len = 4;
  if (buf.length < len) return { needMore: true };
  return { key: null, consume: len };
}

class KeyReader {
  constructor() {
    this.buffer = Buffer.alloc(0);
  }

  readKey() {
    return new Promise((resolve) => {
      const stdin = process.stdin;
      if (!stdin.isTTY) {
        resolve(null);
        return;
      }
      const prevRaw = stdin.isRaw;
      stdin.setRawMode(true);
      stdin.resume();

      const finish = (key) => {
        stdin.setRawMode(prevRaw);
        stdin.removeListener("data", onData);
        resolve(key);
      };

      const drain = () => {
        for (;;) {
          const r = tryParseKey(this.buffer);
          if (r.needMore) return;
          this.buffer = this.buffer.subarray(r.consume);
          finish(r.key);
          return;
        }
      };

      const onData = (chunk) => {
        this.buffer = Buffer.concat([this.buffer, chunk]);
        drain();
      };

      const r0 = tryParseKey(this.buffer);
      if (!r0.needMore) {
        this.buffer = this.buffer.subarray(r0.consume);
        stdin.setRawMode(prevRaw);
        resolve(r0.key);
        return;
      }

      stdin.on("data", onData);
    });
  }
}

function clearScreen() {
  process.stdout.write("\u001b[2J\u001b[H");
}

async function promptTopic() {
  clearScreen();
  process.stdout.write("Topic\n=====\n");
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  try {
    return (await rl.question("Enter topic #: ")).trim();
  } finally {
    rl.close();
  }
}

async function promptWaitInterval() {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  try {
    for (;;) {
      const answer = (await rl.question("Set the interval: ")).trim();
      const interval = parseInt(answer, 10);
      if (interval > 0) return interval;
      process.stdout.write("Interval must be a positive whole number.\n");
    }
  } finally {
    rl.close();
  }
}

function initializeLogFile() {
  const name = `key_presses_${logStamp()}.js`;
  LOG_FILE = path.join(OUTPUT_FOLDER, name);
  fs.mkdirSync(OUTPUT_FOLDER, { recursive: true });
  fs.writeFileSync(LOG_FILE, "export const keyPresses = {\n", "utf8");
}

function logFileDisplayPath() {
  try {
    return path.relative(PROJECT_ROOT, LOG_FILE);
  } catch {
    return LOG_FILE;
  }
}

function finalizeLogFile() {
  fs.appendFileSync(LOG_FILE, "};\n\nexport default keyPresses;\n", "utf8");
}

function startTopic(topic) {
  fs.appendFileSync(
    LOG_FILE,
    `${STEP_INDENT}${JSON.stringify(topic)}: {\n${STEPS_INDENT}steps: [\n`,
    "utf8",
  );
}

function startStep() {
  fs.appendFileSync(LOG_FILE, `${STEP_BLOCK_INDENT}[\n`, "utf8");
}

function logKeyPress(key) {
  fs.appendFileSync(
    LOG_FILE,
    `${ACTION_INDENT}{ type: "remote", key: "KEY_${key}" },\n`,
    "utf8",
  );
}

function logWait(waitMs) {
  fs.appendFileSync(LOG_FILE, `${ACTION_INDENT}{ type: "wait", ms: ${waitMs} },\n`, "utf8");
}

function logCaptureReuse(reuseImage) {
  fs.appendFileSync(
    LOG_FILE,
    `${ACTION_INDENT}{ type: "capture", mode: "reuse", reuseImage: "${reuseImage}" },\n`,
    "utf8",
  );
}

function finishStep() {
  fs.appendFileSync(LOG_FILE, `${STEP_BLOCK_INDENT}],\n`, "utf8");
}

function finishTopic() {
  fs.appendFileSync(LOG_FILE, `${STEPS_INDENT}],\n${STEP_INDENT}}},\n`, "utf8");
}

function lastContentLineIndex(lines) {
  let index = lines.length - 1;
  while (index >= 0 && lines[index] === "") index -= 1;
  return index;
}

function writeLogLines(lines) {
  fs.writeFileSync(LOG_FILE, lines.join("\n"), "utf8");
}

function removeEmptyOpenStep() {
  const lines = fs.readFileSync(LOG_FILE, "utf8").split("\n");
  const index = lastContentLineIndex(lines);
  if (index < 0 || lines[index] !== `${STEP_BLOCK_INDENT}[`) return false;
  lines.splice(index, 1);
  writeLogLines(lines);
  return true;
}

function removeLastStepClose() {
  const lines = fs.readFileSync(LOG_FILE, "utf8").split("\n");
  const index = lastContentLineIndex(lines);
  if (index < 0 || lines[index] !== `${STEP_BLOCK_INDENT}],`) return false;
  lines.splice(index, 1);
  writeLogLines(lines);
  return true;
}

function removeLastActionLine() {
  const lines = fs.readFileSync(LOG_FILE, "utf8").split("\n");
  for (let index = lastContentLineIndex(lines); index >= 0; index -= 1) {
    if (lines[index].startsWith(ACTION_INDENT)) {
      lines.splice(index, 1);
      writeLogLines(lines);
      return true;
    }
    if (lines[index] === `${STEP_BLOCK_INDENT}[`) break;
  }
  return false;
}

function findLastRemoteKey(actions) {
  for (let i = actions.length - 1; i >= 0; i -= 1) {
    const [actionType, value] = actions[i];
    if (actionType === "remote") return String(value);
  }
  return null;
}

function recomputeCounts(counts, actions) {
  for (const k of Object.keys(counts)) delete counts[k];
  for (const [actionType, value] of actions) {
    if (actionType === "remote") counts[String(value)] = (counts[String(value)] ?? 0) + 1;
  }
  return findLastRemoteKey(actions);
}

function describeAction(action) {
  const [actionType, value] = action;
  if (actionType === "wait") return `Wait ${value} ms`;
  if (actionType === "capture") return `Capture reuse ${value}`;
  return KEY_NAMES[String(value)];
}

function formatTime(d) {
  const z2 = (n) => String(n).padStart(2, "0");
  return `${z2(d.getHours())}:${z2(d.getMinutes())}:${z2(d.getSeconds())}`;
}

function draw(counts, recent, lastKey, topicOpen, stepOpen, currentTopic, waitMs) {
  clearScreen();
  process.stdout.write(
    [
      "W to add a wait",
      "F to finish current step.",
      "I to set wait interval.",
      "T to start the next topic.",
      "P to capture previous image.",
      'R to reuse images "1-3".',
      "Backspace to undo the last action.",
      "Q to end program.\n",
      `Saving presses to: ${logFileDisplayPath()}`,
      `Wait interval: ${waitMs} ms`,
    ].join("\n") + "\n",
  );
  if (topicOpen && stepOpen && currentTopic) {
    process.stdout.write(`Current topic: ${currentTopic}\n\n`);
  } else if (currentTopic) {
    process.stdout.write(`Current topic: ${currentTopic} (ready)\n\n`);
  } else {
    process.stdout.write("Current topic: not started\n\n");
  }

  for (const key of ["UP", "DOWN", "LEFT", "RIGHT", "HOME", "ENTER"]) {
    process.stdout.write(`${KEY_NAMES[key]}: ${counts[key] ?? 0}\n`);
  }

  if (lastKey) process.stdout.write(`\nLast key: ${KEY_NAMES[lastKey]}\n`);
  else process.stdout.write("\nLast key: none yet\n");

  if (recent.length) {
    process.stdout.write("\nKey Counter:\n");
    for (const event of recent) process.stdout.write(`  ${event}\n`);
  }
}

function suspendSigint() {
  const prev = [...process.listeners("SIGINT")];
  process.removeAllListeners("SIGINT");
  process.on("SIGINT", () => {});
  return () => {
    process.removeAllListeners("SIGINT");
    for (const l of prev) process.on("SIGINT", l);
  };
}

async function main() {
  if (!process.stdin.isTTY || !process.stdout.isTTY) {
    process.stderr.write("keyTracker requires an interactive terminal (TTY).\n");
    process.exitCode = 1;
    return;
  }

  const restoreSigint = suspendSigint();
  const reader = new KeyReader();
  /** @type {Record<string, number>} */
  const counts = {};
  /** @type {string[]} */
  const recent = [];
  const RECENT_MAX = 3;

  function pushRecent(line) {
    recent.unshift(line);
    while (recent.length > RECENT_MAX) recent.pop();
  }

  let lastKey = null;
  let topicOpen = false;
  let stepOpen = false;
  let currentTopic = null;
  /** @type {Action[]} */
  let currentStepActions = [];
  /** @type {Action[][]} */
  const finishedStepActions = [];
  let waitMs = DEFAULT_WAIT_MS;

  async function openStepIfNeeded() {
    if (stepOpen) return;
    if (currentTopic === null) {
      currentTopic = await promptTopic();
    }
    if (!topicOpen) {
      startTopic(currentTopic);
      topicOpen = true;
    }
    startStep();
    currentStepActions = [];
    stepOpen = true;
  }

  function redraw() {
    draw(counts, recent, lastKey, topicOpen, stepOpen, currentTopic, waitMs);
  }

  initializeLogFile();

  try {
    currentTopic = await promptTopic();
    redraw();

    for (;;) {
      const key = await reader.readKey();

      if (key === "QUIT") break;
      const timestamp = new Date();

      if (key === "SET_INTERVAL") {
        waitMs = await promptWaitInterval();
        pushRecent(`${formatTime(timestamp)} - Wait interval ${waitMs} ms`);
        redraw();
        continue;
      }

      if (key === "WAIT") {
        await openStepIfNeeded();
        logWait(waitMs);
        currentStepActions.push(["wait", waitMs]);
        finishedStepActions.push([...currentStepActions]);
        finishStep();
        currentStepActions = [];
        stepOpen = false;
        for (const k of Object.keys(counts)) delete counts[k];
        lastKey = null;
        pushRecent(`${formatTime(timestamp)} - Wait ${waitMs} ms; finished step`);
        redraw();
        continue;
      }

      if (key === "FINISH") {
        if (stepOpen) {
          if (currentStepActions.length) {
            finishedStepActions.push([...currentStepActions]);
            finishStep();
          } else {
            removeEmptyOpenStep();
          }
          currentStepActions = [];
          stepOpen = false;
          pushRecent(`${formatTime(timestamp)} - Finished step`);
          for (const k of Object.keys(counts)) delete counts[k];
          lastKey = null;
        } else {
          pushRecent(`${formatTime(timestamp)} - No open step to finish`);
        }
        redraw();
        continue;
      }

      if (key === "UNDO") {
        if (!stepOpen && topicOpen && finishedStepActions.length) {
          if (removeLastStepClose()) {
            currentStepActions = finishedStepActions.pop();
            lastKey = recomputeCounts(counts, currentStepActions);
            stepOpen = true;
          } else {
            pushRecent(`${formatTime(timestamp)} - Could not go back to step`);
            redraw();
            continue;
          }
        }

        if (!stepOpen || !currentStepActions.length) {
          pushRecent(`${formatTime(timestamp)} - Nothing to undo`);
          redraw();
          continue;
        }

        const action = currentStepActions.pop();
        if (!removeLastActionLine()) {
          currentStepActions.push(action);
          pushRecent(`${formatTime(timestamp)} - Could not undo last action`);
          redraw();
          continue;
        }

        if (action[0] === "remote") {
          const vk = String(action[1]);
          counts[vk] = Math.max(0, (counts[vk] ?? 0) - 1);
        }

        if (currentStepActions.length) {
          lastKey = findLastRemoteKey(currentStepActions);
        } else {
          removeEmptyOpenStep();
          stepOpen = false;
          for (const k of Object.keys(counts)) delete counts[k];
          lastKey = null;
        }

        pushRecent(`${formatTime(timestamp)} - Removed ${describeAction(action)}`);
        redraw();
        continue;
      }

      if (key === "NEXT_TOPIC") {
        if (stepOpen) {
          if (currentStepActions.length) {
            finishedStepActions.push([...currentStepActions]);
            finishStep();
          } else {
            removeEmptyOpenStep();
          }
          currentStepActions = [];
          stepOpen = false;
        }
        if (topicOpen) {
          finishTopic();
          topicOpen = false;
        }

        for (const k of Object.keys(counts)) delete counts[k];
        lastKey = null;
        finishedStepActions.length = 0;
        currentTopic = await promptTopic();
        pushRecent(`${formatTime(new Date())} - Ready for topic ${currentTopic}`);
        redraw();
        continue;
      }

      if (key != null && Object.prototype.hasOwnProperty.call(CAPTURE_REUSE_IMAGES, key)) {
        const reuseImage = CAPTURE_REUSE_IMAGES[key];
        await openStepIfNeeded();
        logCaptureReuse(reuseImage);
        currentStepActions.push(["capture", reuseImage]);
        pushRecent(`${formatTime(timestamp)} - Capture reuse ${reuseImage}`);
        redraw();
        continue;
      }

      if (key == null || !Object.prototype.hasOwnProperty.call(KEY_NAMES, key)) continue;

      counts[key] = (counts[key] ?? 0) + 1;
      lastKey = key;
      await openStepIfNeeded();
      logKeyPress(key);
      currentStepActions.push(["remote", key]);
      pushRecent(`${formatTime(timestamp)} - ${KEY_NAMES[key]}`);
      redraw();
    }
  } catch {
    // mirror Python KeyboardInterrupt pass
  } finally {
    restoreSigint();
    if (stepOpen) {
      if (currentStepActions.length) finishStep();
      else removeEmptyOpenStep();
    }
    if (topicOpen) finishTopic();
    finalizeLogFile();
  }
}

await main();
